src/ucp/rndv/rndv_ppln.c

```diff
@@ -85,52 +85,81 @@ ucp_proto_rndv_ppln_probe(const ucp_proto_init_params_t *init_params)
+        rpriv.frag_proto_min_length = ucp_proto_perf_segment_start(first_seg);
+        ucs_assertv(rpriv.frag_size >= rpriv.frag_proto_min_length,
+                    "rpriv.frag_size=%zu rpriv.frag_proto_min_length=%zu",
+                    rpriv.frag_size, rpriv.frag_proto_min_length);
 
         frag_proto_priv = &ucs_array_elem(&select_elem->proto_init.priv_buf,
                                           proto->priv_offset);
         ucp_proto_rndv_set_variant_config(init_params, proto,
                                           &sel_param, frag_proto_priv,
                                           &rpriv.frag_proto_cfg);
 
-        ucs_trace("rndv_ppln frag %s" UCP_PROTO_PERF_FUNC_TYPES_FMT,
+        ucp_proto_perf_segment_str(frag_seg, &seg_strb);
+        ucs_trace("rndv_ppln frag: %s proto: %s segment: %s",
                   ucs_memunits_to_str(rpriv.frag_size, frag_size_str,
                                       sizeof(frag_size_str)),
-                  UCP_PROTO_PERF_FUNC_TYPES_ARG(frag_range->perf));
+                  ucp_proto_id_field(proto->proto_id, name),
+                  ucs_string_buffer_cstr(&seg_strb));
 
         /* Add the single range of the pipeline protocol to ppln_caps */
-        ppln_caps.min_length = frag_range->max_length + 1;
-        ppln_caps.num_ranges = 0;
-        ucp_proto_common_add_ppln_range(&ppln_caps, frag_range, SIZE_MAX);
+        status = ucp_proto_perf_create("pipeline", &ppln_perf);
+        if (status != UCS_OK) {
+            continue;
+        }
+
+        status = ucp_perf_add_ppln(ppln_perf, frag_seg, SIZE_MAX);
+        if (status != UCS_OK) {
+            goto out_destroy_ppln_perf;
+        }
 
         /* Add ATS overhead */
-        ppln_overhead = ucs_linear_func_make(
-                frag_overhead, frag_overhead / frag_range->max_length);
-        status        = ucp_proto_rndv_ack_init(
-                init_params, UCP_PROTO_RNDV_ATS_NAME, &ppln_caps,
-                ppln_overhead, &rpriv.ack, &caps);
-        if (status == UCS_OK) {
-            ucp_proto_select_add_proto(init_params, proto->cfg_thresh,
-                                       proto->cfg_priority,
-                                       &caps, &rpriv, sizeof(rpriv));
+        ppln_overhead = ucs_linear_func_make(frag_overhead,
+                                             frag_overhead / rpriv.frag_size);
+        status        = ucp_proto_rndv_ack_init(&ack_params,
+                                                UCP_PROTO_RNDV_ATS_NAME,
+                                                ppln_overhead, &ack_perf,
+                                                &rpriv.ack);
+        if ((status != UCS_OK) || (rpriv.ack.lane == UCP_NULL_LANE)) {
+            goto out_destroy_ppln_perf;
         }
-        ucp_proto_select_caps_cleanup(&ppln_caps);
+
+        status = ucp_proto_perf_aggregate2(
+                ucp_proto_id_field(init_params->proto_id, name),
```

Comment:
shall we add local var proto_name? it is used in several places in this function

Reply:
I see that it used only once for `init_params->proto_id` and once for `proto->proto_id`